Model/Currency.php

```diff
@@ -235,18 +245,60 @@ protected function checkLocalIP($ip)
+     * then it is added in the cache (magento style of caching without using third party software component
+     * like redis, varnish)
+     * @param string $currencyCode
      * @param float $amount
      * @return int|float
      */
-    public function convertCurrency($currencycode,$amount)
+    public function convertCurrency($currencyCode, $amount)
     {
-        if($currencycode == "JPY")
-        {
-            return round($amount);
+        $precision_adjusted = 0;
+        $precision = 0;
+        if (isset($currencyCode)) {
+            $reachCache = $this->cacher->loadDataFromCache();
+            $this->_logger->debug("this is in cache :::" . json_encode($reachCache));
+            if (isset($reachCache[$currencyCode])) {
+                $this->_logger->debug("Loaded Currency Precision from Cache :::" . json_encode($reachCache));
+                $precision = $reachCache[$currencyCode];
+            }
+            else {
+                $data = $this->getResource()->getPrecisionByCurrency($currencyCode);
+                $this->_logger->debug("Currency info from DB :::" . json_encode($data));
+                if (isset($data[$currencyCode])) {
+                    $precision = $data[$currencyCode]['precision_unit'];
+                    $this->_logger->debug("Getting precision from Database call :" . $precision);
+                } else {
+                    $this->_logger->debug("Getting precision from API call");
+                    $url = $this->reachHelper->getApiUrl();
+                    $this->_logger->debug(json_encode($url));
+                    $url .= 'localize?MerchantId=' . $this->reachHelper->getMerchantId() . "&Currency=" . $currencyCode;
+                    $this->_logger->debug("url to retreive currency precision: " . json_encode($url));
+                    $rest = $this->httpRestFactory->create();
+                    $rest->setUrl($url);
+                    $response = $rest->executeGet();
+                    $result = $response->getResponseData();
+                    if (isset($result)) {
+                        $precision = $result['Units'];
+                        $this->getResource()->setPrecisionByCurrency($currencyCode, $precision);
+                    }
+                }
+                $currencyVsPrecision = [$currencyCode => $precision];
+                $this->cacher->saveDataInCache($currencyVsPrecision);
+            }
+            $precision_adjusted = ($precision > self::PRECISION_CUTOFF)? self::PRECISION_CUTOFF: $precision; //because of a bug in
+            //our API (as per the note in description of JIRA MAG-100)
+            $this->_logger->debug("in convert currency: precision adjusted :".$precision_adjusted);
+
         }
-        return round($amount, self::PRECISION);
+        return round($amount, $precision_adjusted); //when nothing coming back from api or db; verify with team
```

Comment:
ideally in the unlikely event that you don't get anything back from the db, cache or Reach API I can see two approaches:

- default to 0 precision because that will work with all currencies, and the merchant/us will likely notice it faster
- default to 2 precision because that matches what most currencies need anyway, but anything with 0 precision (JPY) will fail across the board

Reply:
the safer choice that retains customer's ability to checkout would be defaulting to 0, so I'm leaning that way